packages/flutter/lib/src/services/text_input.dart

```diff
@@ -671,6 +686,370 @@ class RawFloatingCursorPoint {
+  /// that was deleted.
+  ///
+  /// For a [TextEditingDeltaType.replacement] this will be the range of
+  /// characters that are being replaced.
+  ///
+  /// For a [TextEditingDeltaType.equality] this will be a collapsed range
+  /// of (-1,-1).
+  final TextRange deltaRange;
+
+  /// {@template flutter.services.TextEditingDelta.deltaType}
+  /// The type of delta that has occured.
+  ///
+  /// See [TextEditingDeltaType] for more information.
+  /// {@endtemplate}
+  TextEditingDeltaType get deltaType;
+
+  /// The range of text that is currently selected after the delta has been
+  /// applied.
+  final TextSelection selection;
+
+  /// The range of text that is still being composed after the delta has been
+  /// applied.
+  final TextRange composing;
+
+  /// {@template flutter.services.TextEditingDelta.apply}
+  /// This method will take the given [TextEditingValue] and return a new
+  /// [TextEditingValue] with that instance of [TextEditingDelta] applied to it.
+  /// {@endtemplate}
+  TextEditingValue apply(TextEditingValue value);
+}
+
+/// {@macro flutter.services.TextEditingDeltaInsertion}
+class TextEditingDeltaInsertion extends TextEditingDelta {
+  /// Creates an insertion delta for a given change to the editing state.
+  ///
+  /// {@macro flutter.services.TextEditingDelta}
+  const TextEditingDeltaInsertion({
+    required String oldText,
+    required String deltaText,
+    required TextRange deltaRange,
+    required TextSelection selection,
+    required TextRange composing,
+  }) : super(
+      oldText: oldText,
+      deltaText: deltaText,
+      deltaRange: deltaRange,
+      selection: selection,
+      composing:composing,
+  );
+
+  /// Creates an instance of this class from a JSON object.
+  factory TextEditingDeltaInsertion.fromJSON(Map<String, dynamic> encoded) {
+    return TextEditingDeltaInsertion(
+      oldText: encoded['oldText'] as String,
+      deltaText: encoded['deltaText'] as String,
+      deltaRange: TextRange(
+        start: encoded['deltaStart'] as int? ?? -1,
+        end: encoded['deltaEnd'] as int? ?? -1,
+      ),
+      selection: TextSelection(
+        baseOffset: encoded['selectionBase'] as int? ?? -1,
+        extentOffset: encoded['selectionExtent'] as int? ?? -1,
+        affinity: _toTextAffinity(encoded['selectionAffinity'] as String?) ??
+            TextAffinity.downstream,
+        isDirectional: encoded['selectionIsDirectional'] as bool? ?? false,
+      ),
+      composing: TextRange(
+        start: encoded['composingBase'] as int? ?? -1,
+        end: encoded['composingExtent'] as int? ?? -1,
+      ),
+    );
+  }
+
+  /// {@macro flutter.services.TextEditingDelta.deltaType}
+  @override
+  TextEditingDeltaType get deltaType => TextEditingDeltaType.insertion;
+
+  /// {@macro flutter.services.TextEditingDelta.apply}
+  @override
+  TextEditingValue apply(TextEditingValue value) {
+    String newText = value.text;
```

Comment:
Should this be using `oldText` instead of `value.text`? It may crash if `value.text` is, say, empty.

Reply:
I have added a check to verify if `oldText == value.text`. If this condition is not true then we return the original `TextEditingValue`.